### PR TITLE
r/glue_partition - add new resource

### LINKED ISSUE #3878: Feature Request: Resource for managing AWS Glue Partitions
Split feature request from #1416. For issues not immediately being worked on, please use :+1: upvotes on this original issue comment to help guage community interest.

### Terraform Version

```
terraform 0.10+
terraform-provider-aws (new feature)
```

### Affected Resource(s)

- `aws_glue_partition`

### Terraform Configuration Files

```hcl
# New Resource - not implemented
resource "aws_glue_partition" "example" {
  # ...
}
```

### Expected Behavior

Create, read, update, and delete support for AWS Glue partitions.

### Actual Behavior

Resource not implemented.

### References

* https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api-catalog-partitions.html
* https://docs.aws.amazon.com/sdk-for-go/api/service/glue/#Glue.CreatePartition

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #3878

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource_aws_glue_partition: add new resource
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSGluePartition_'
--- PASS: TestAccAWSGluePartition_basic (65.18s)
--- PASS: TestAccAWSGluePartition_parameters (166.56s)
--- PASS: TestAccAWSGluePartition_disappears (52.11s)
--- PASS: TestAccAWSGluePartition_multipleValues (66.00s)
```

### CODE REVIEW

File: aws/resource_aws_glue_partition.go [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+		PartitionInput: expandGluePartitionInput(d),
+	}
+
+	log.Printf("[DEBUG] Creating Glue Partition: %#v", input)
+	_, err := conn.CreatePartition(input)
+	if err != nil {
+		return fmt.Errorf("error creating Glue Partition: %w", err)
+	}
+
+	d.SetId(fmt.Sprintf("%s:%s:%s:%s", catalogID, dbName, tableName, stringifyAwsGluePartition(values)))
```
Comment (ewbankkit) [inline aws/resource_aws_glue_partition.go:226]: Maybe add a _create Glue partition ID_ function to parallel `readAwsGluePartitionID`.
There is also now precedent in moving these ID create/read functions to an `internal/service` package.
See e.g. `aws/internal/service/ec2/id.go`.

**Review by ewbankkit — APPROVED** [review]
LGTM.

```console
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSGluePartition_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSGluePartition_ -timeout 120m
=== RUN   TestAccAWSGluePartition_basic
=== PAUSE TestAccAWSGluePartition_basic
=== RUN   TestAccAWSGluePartition_disappears
=== PAUSE TestAccAWSGluePartition_disappears
=== CONT  TestAccAWSGluePartition_basic
=== CONT  TestAccAWSGluePartition_disappears
    resource_aws_glue_partition_test.go:48: [INFO] Got non-empty plan, as expected
--- PASS: TestAccAWSGluePartition_disappears (19.76s)
--- PASS: TestAccAWSGluePartition_basic (22.14s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	22.181s
```

File: aws/resource_aws_glue_partition.go (hunk 1 above)
Comment (DrFaust92) [reply aws/resource_aws_glue_partition.go]: done.

**Comment by DrFaust92** [comment]
Rebased and revised tests a bit:

```
--- PASS: TestAccAWSGluePartition_basic (65.18s)
--- PASS: TestAccAWSGluePartition_parameters (166.56s)
--- PASS: TestAccAWSGluePartition_disappears (52.11s)
--- PASS: TestAccAWSGluePartition_multipleValues (66.00s)
```

**Review by ewbankkit — APPROVED** [review]
LGTM.

```console
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSGluePartition_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSGluePartition_ -timeout 120m
=== RUN   TestAccAWSGluePartition_basic
=== PAUSE TestAccAWSGluePartition_basic
=== RUN   TestAccAWSGluePartition_multipleValues
=== PAUSE TestAccAWSGluePartition_multipleValues
=== RUN   TestAccAWSGluePartition_parameters
=== PAUSE TestAccAWSGluePartition_parameters
=== RUN   TestAccAWSGluePartition_disappears
=== PAUSE TestAccAWSGluePartition_disappears
=== CONT  TestAccAWSGluePartition_basic
=== CONT  TestAccAWSGluePartition_disappears
=== CONT  TestAccAWSGluePartition_parameters
=== CONT  TestAccAWSGluePartition_multipleValues
=== CONT  TestAccAWSGluePartition_disappears
    resource_aws_glue_partition_test.go:126: [INFO] Got non-empty plan, as expected
--- PASS: TestAccAWSGluePartition_disappears (26.79s)
--- PASS: TestAccAWSGluePartition_basic (30.38s)
--- PASS: TestAccAWSGluePartition_multipleValues (30.49s)
--- PASS: TestAccAWSGluePartition_parameters (59.75s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	59.790s
```

**Comment by breathingdust** [comment]
LGTM! 🚀  Thanks @DrFaust92 

Verified Acceptance Tests in Commercial (us-west-2)

```console
make testacc TEST=./aws TESTARGS='-run=TestAccAWSGluePartition_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSGluePartition_ -timeout 120m
=== RUN   TestAccAWSGluePartition_basic
=== PAUSE TestAccAWSGluePartition_basic
=== RUN   TestAccAWSGluePartition_multipleValues
=== PAUSE TestAccAWSGluePartition_multipleValues
=== RUN   TestAccAWSGluePartition_parameters
=== PAUSE TestAccAWSGluePartition_parameters
=== RUN   TestAccAWSGluePartition_disappears
=== PAUSE TestAccAWSGluePartition_disappears
=== CONT  TestAccAWSGluePartition_basic
=== CONT  TestAccAWSGluePartition_disappears
=== CONT  TestAccAWSGluePartition_multipleValues
=== CONT  TestAccAWSGluePartition_parameters
2020/10/01 11:46:14 [WARN] Truncating attribute path of 0 diagnostics for TypeSet
=== CONT  TestAccAWSGluePartition_disappears
    resource_aws_glue_partition_test.go:126: [INFO] Got non-empty plan, as expected
--- PASS: TestAccAWSGluePartition_disappears (16.69s)
--- PASS: TestAccAWSGluePartition_multipleValues (18.92s)
--- PASS: TestAccAWSGluePartition_basic (18.93s)
--- PASS: TestAccAWSGluePartition_parameters (43.35s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	45.203s
```

Verified Acceptance Tests in GovCloud (us-gov-west-1)

```console
make testacc TEST=./aws TESTARGS='-run=TestAccAWSGluePartition_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSGluePartition_ -timeout 120m
=== RUN   TestAccAWSGluePartition_basic
=== PAUSE TestAccAWSGluePartition_basic
=== RUN   TestAccAWSGluePartition_multipleValues
=== PAUSE TestAccAWSGluePartition_multipleValues
=== RUN   TestAccAWSGluePartition_parameters
=== PAUSE TestAccAWSGluePartition_parameters
=== RUN   TestAccAWSGluePartition_disappears
=== PAUSE TestAccAWSGluePartition_disappears
=== CONT  TestAccAWSGluePartition_basic
=== CONT  TestAccAWSGluePartition_disappears
=== CONT  TestAccAWSGluePartition_multipleValues
=== CONT  TestAccAWSGluePartition_parameters
=== CONT  TestAccAWSGluePartition_disappears
    resource_aws_glue_partition_test.go:126: [INFO] Got non-empty plan, as expected
--- PASS: TestAccAWSGluePartition_disappears (15.30s)
--- PASS: TestAccAWSGluePartition_basic (15.83s)
--- PASS: TestAccAWSGluePartition_multipleValues (15.97s)
--- PASS: TestAccAWSGluePartition_parameters (34.31s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	35.986s
```

**Pull request merged**